platform/darwin/src/MGLGeoJSONSource.mm

```diff
@@ -7,25 +7,113 @@
+NSString * const MGLGeoJSONBufferOptionName = @"MGLGeoJSONBufferOptionName";
+NSString * const MGLGeoJSONToleranceOptionName = @"MGLGeoJSONOptionsClusterTolerance";
+
+@interface MGLGeoJSONSource ()
+
+@property (nonatomic, readwrite) NSDictionary *options;
+
+@end
+
 @implementation MGLGeoJSONSource
 
-- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier geoJSONData:(NSData *)data {
-    if (self = [super initWithSourceIdentifier:sourceIdentifier]) {
+- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier geoJSONData:(NSData *)data
+{
+    if (self = [super initWithSourceIdentifier:sourceIdentifier])
+    {
         _geoJSONData = data;
     }
     return self;
 }
 
-- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier URL:(NSURL *)url {
-    if (self = [super initWithSourceIdentifier:sourceIdentifier]) {
+- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier geoJSONData:(NSData *)data options:(NS_DICTIONARY_OF(NSString *, id) *)options
+{
+    if (self = [super initWithSourceIdentifier:sourceIdentifier])
+    {
+        _geoJSONData = data;
+        _options = options;
+    }
+    return self;
+}
+
+- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier URL:(NSURL *)url
+{
+    if (self = [super initWithSourceIdentifier:sourceIdentifier])
+    {
+        _URL = url;
+    }
+    return self;
+}
+
+- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier URL:(NSURL *)url options:(NS_DICTIONARY_OF(NSString *, id) *)options
+{
+    if (self = [super initWithSourceIdentifier:sourceIdentifier])
+    {
         _URL = url;
+        _options = options;
     }
     return self;
 }
 
+- (mbgl::style::GeoJSONOptions)mbgl_geoJSONOptions
+{
+    auto options = mbgl::style::GeoJSONOptions();
+    
+    setOptionForKeyWithValue(MGLGeoJSONMaxZoomLevelOptionName, &options.maxzoom, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONBufferOptionName, &options.buffer, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONToleranceOptionName, &options.tolerance, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONClusterRadiusOptionName, &options.clusterRadius, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONClusterMaximumZoomLevelOptionName, &options.clusterMaxZoom, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONClusterOptionName, &options.cluster, self.options);
+    
+    return options;
+}
+
+template<typename T>
+void setOptionForKeyWithValue(NSString *key, T *value, NSDictionary *options)
+{
+    if (! [options objectForKey:key]) return;
+   
+    id keyValue = [options objectForKey:key];
+    BOOL isKindOfNumber = [keyValue isKindOfClass:[NSNumber class]];
+    
+    // if we are dealing with an NSNumber then determine the selector to extract
+    // its value based on its type and then call that selector, assigning the value
+    // to the passed in mbgl options value
+    if (isKindOfNumber)
```

Comment:
Sorry, you’re right, I was thinking of the opposite direction.

Looking at this again, I don’t think this method should even exist. For example, we document that `MGLGeoJSONToleranceOptionName` is an NSNumber containing a double. So we should just set `options.tolerance` to `[options[@"MGLGeoJSONToleranceOptionName"] doubleValue]` and possibly assert if it isn’t an NSNumber. That way, if the developer passes in an NSNumber containing an integer, that value is automatically converted to a double.